File: src/main/java/com/sequenceiq/cloudbreak/repository/StackRepository.java
package com.sequenceiq.cloudbreak.repository;

import java.util.List;
import java.util.Set;

import org.springframework.data.repository.CrudRepository;
import org.springframework.data.repository.query.Param;
import org.springframework.security.access.prepost.PostAuthorize;

import com.sequenceiq.cloudbreak.domain.Stack;

public interface StackRepository extends CrudRepository<Stack, Long> {

    @PostAuthorize("hasPermission(returnObject,'read')")
    Stack findOne(@Param("id") Long id);

    Stack findById(@Param("id") Long id);

    Stack findByAmbari(@Param("ambariIp") String ambariIp);

    Set<Stack> findForUser(@Param("user") String user);

    Set<Stack> findPublicsInAccount(@Param("account") String account);

    Set<Stack> findAllInAccount(@Param("account") String account);

    Stack findOneWithLists(@Param("id") Long id);

    Stack findByStackResourceName(@Param("stackName") String stackName);

    List<Stack> findAllStackForTemplate(@Param("id") Long id);

    Stack findStackForCluster(@Param("id") Long id);

    Stack findStackWithListsForCluster(@Param("id") Long id);

    List<Stack> findRequestedStacksWithCredential(@Param("credentialId") Long credentialId);

    Stack findStackByHash(@Param("hash") String hash);

    Stack findByNameInAccount(@Param("name") String name, @Param("account") String account, @Param("owner") String owner);

    Stack findByNameInUser(@Param("name") String name, @Param("owner") String owner);

    @PostAuthorize("hasPermission(returnObject,'read')")
    Stack findOneByName(@Param("name") String name);

}

File: core/src/main/java/com/sequenceiq/cloudbreak/repository/StackRepository.kt
package com.sequenceiq.cloudbreak.repository

import org.springframework.data.repository.CrudRepository
import org.springframework.data.repository.query.Param

import com.sequenceiq.cloudbreak.api.model.Status
import com.sequenceiq.cloudbreak.domain.Stack

@EntityType(entityClass = Stack::class)
interface StackRepository : CrudRepository<Stack, Long> {

    override fun findOne(@Param("id") id: Long?): Stack

    fun findById(@Param("id") id: Long?): Stack

    fun findByIdLazy(@Param("id") id: Long?): Stack

    fun findByAmbari(@Param("ambariIp") ambariIp: String): Stack

    fun findForUser(@Param("user") user: String): Set<Stack>

    fun findPublicInAccountForUser(@Param("user") user: String, @Param("account") account: String): Set<Stack>

    fun findAllInAccount(@Param("account") account: String): Set<Stack>

    fun findOneWithLists(@Param("id") id: Long?): Stack

    fun findAllStackForTemplate(@Param("id") id: Long?): List<Stack>

    fun findStackForCluster(@Param("id") id: Long?): Stack

    fun findByIdInAccount(@Param("id") id: Long?, @Param("account") account: String): Stack

    fun findByNameInAccount(@Param("name") name: String, @Param("account") account: String, @Param("owner") owner: String): Stack

    fun findByNameInUser(@Param("name") name: String, @Param("owner") owner: String): Stack

    fun findOneByName(@Param("name") name: String, @Param("account") account: String): Stack

    fun findByCredential(@Param("credentialId") credentialId: Long?): List<Stack>

    fun findAllByNetwork(@Param("networkId") networkId: Long?): List<Stack>

    fun findByIdWithSecurityConfig(@Param("id") id: Long?): Stack

    fun findByIdWithSecurityGroup(@Param("id") id: Long?): Stack

    fun findAllBySecurityGroup(@Param("securityGroupId") securityGroupId: Long?): List<Stack>

    fun findAllAlive(): List<Stack>

    fun findByStatuses(@Param("statuses") statuses: List<Status>): List<Stack>

    fun findStacksWithoutEvents(): Set<Long>
}
